src/originWhitelist.js

```diff
@@ -8,46 +8,71 @@
  */
 const {URL} = require('url');
 
+// Match origins:
+// https://studio.code.org                          Production
+// https://test-studio.code.org                     Test
+// https://dashboard-adhoc-my-branch.cdn-code.org   Ad-Hoc servers
+// http://localhost-studio.code.org:3000            Local development
+const CODE_ORG_URL = /^https?:\/\/(?:[\w\d-]+\.)?(?:cdn-)?code\.org(?::\d+)?$/i;
+
 /**
  * Navigation to urls matching any of the given origins will open the page in
  * the system default browser, instead of within Maker Toolkit Browser, even
  * if the origin falls within our general Code.org origin whitelist.
  * @type {Array.<RegExp>}
  */
-const BLACKLIST = [
+const INTERNAL_BLACKLIST = [
   /curriculum\.code\.org$/i,
   /docs\.code\.org$/i,
   /forum\.code\.org$/i,
   /support\.code\.org$/i,
   /wiki\.code\.org$/i,
 ];
 
-// Match origins:
-// https://studio.code.org                          Production
-// https://test-studio.code.org                     Test
-// https://dashboard-adhoc-my-branch.cdn-code.org   Ad-Hoc servers
-// http://localhost-studio.code.org:3000            Local development
-const CODE_ORG_URL = /^https?:\/\/(?:[\w\d-]+\.)?(?:cdn-)?code\.org(?::\d+)?$/i;
-
 /**
- * @param {string} origin
- * @returns {boolean} whether the origin is included in the whitelist
+ * Limited set of non-Code.org pages we will load within Maker Toolkit Browser,
+ * mostly for things like OAuth or other integrations with external services.
+ * We won't inject native APIs to these sites even though they load within the
+ * app.
+ * @type {Array.<RegExp>}
  */
-function isOriginWhitelisted(origin) {
-  return CODE_ORG_URL.test(origin) &&
-    !BLACKLIST.some(site => site.test(origin));
-}
+const EXTERNAL_WHITELIST = [
+  // Google OAuth
+  /accounts\.google\.com(?:$|\/)/i,
+  // Facebook OAuth
+  /www\.facebook\.com\/v2.6\/dialog\/oauth/i,
```

Comment:
Is v2.6 going to be frozen somehow, or is that something that may change at Facebook's discretion?

Reply:
It's [baked into the Code.org Facebook OAuth configuration here](https://github.com/code-dot-org/code-dot-org/blob/0b18f09a6f90dfdd8ed5afbcf61d0cb329d3f732/dashboard/config/initializers/devise.rb#L278-L280) so if it changes out from under us, at least it's our own fault.  I considered whitelisting something a little broader, like `/www\.facebook\.com\/.*\/oauth/i` but even that's not a guarantee if we switch to a new API version.